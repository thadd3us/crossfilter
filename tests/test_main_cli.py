"""Tests for the main CLI application."""

import socket
import subprocess
import time
from contextlib import closing

import pytest
import requests


def find_free_port() -> int:
    """Find an available port on localhost."""
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
        s.bind(('', 0))
        s.listen(1)
        port = s.getsockname()[1]
    return port


def wait_for_server(host: str, port: int, timeout: float = 10.0) -> bool:
    """Wait for server to start accepting connections."""
    start_time = time.time()
    while time.time() - start_time < timeout:
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(1)
                result = sock.connect_ex((host, port))
                if result == 0:
                    return True
        except Exception:
            pass
        time.sleep(0.1)
    return False


def test_cli_server_startup_and_html_content() -> None:
    """Test that the CLI starts the server and serves HTML with 'Crossfilter' in title."""
    port = find_free_port()
    host = "127.0.0.1"
    
    # Start the server process
    process = subprocess.Popen([
        "uv", "run", "python", "-m", "gpx_viewer.main", "serve", 
        "--port", str(port), "--host", host
    ], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    
    try:
        # Wait for server to start
        server_started = wait_for_server(host, port, timeout=10.0)
        if not server_started:
            process.terminate()
            process.wait()
            stderr = process.stderr.read().decode('utf-8') if process.stderr else ''
            raise AssertionError(f"Server failed to start on {host}:{port}, {stderr=}")
        
        # Give server a moment to fully initialize
        time.sleep(0.5)
        
        # Make request to the main page
        response = requests.get(f"http://{host}:{port}/", timeout=5)
        assert response.status_code == 200
        
        # Check that the HTML contains "Crossfilter" in the title
        html_content = response.text
        assert "Crossfilter" in html_content, f"'Crossfilter' not found in HTML content: {html_content}"
        
        # Additional checks for expected content
        assert "<title>" in html_content
        assert "Crossfilter" in html_content
        
    finally:
        # Clean up: terminate the server process
        process.terminate()
        try:
            process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            process.kill()
            process.wait()


def test_cli_help_command():
    """Test that the CLI help command works."""
    result = subprocess.run([
        "uv", "run", "python", "-m", "gpx_viewer.main", "--help"
    ], capture_output=True, text=True)
    
    assert result.returncode == 0
    assert "Crossfilter" in result.stdout
    assert "serve" in result.stdout


def test_serve_command_help():
    """Test that the serve command help works."""
    result = subprocess.run([
        "uv", "run", "python", "-m", "gpx_viewer.main", "serve", "--help"
    ], capture_output=True, text=True)
    
    assert result.returncode == 0
    assert "--port" in result.stdout
    assert "--host" in result.stdout