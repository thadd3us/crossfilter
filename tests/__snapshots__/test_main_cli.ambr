# serializer version: 1
# name: test_cli_server_with_preload_jsonl
  dict({
    'all_rows_count': 100,
    'columns': list([
      'UUID_STRING',
      'DATA_TYPE',
      'NAME',
      'CAPTION',
      'SOURCE_FILE',
      'TIMESTAMP_MAYBE_TIMEZONE_AWARE',
      'TIMESTAMP_UTC',
      'GPS_LATITUDE',
      'GPS_LONGITUDE',
      'RATING_0_TO_5',
      'SIZE_IN_BYTES',
      'QUANTIZED_H3_L0',
      'QUANTIZED_H3_L1',
      'QUANTIZED_H3_L2',
      'QUANTIZED_H3_L3',
      'QUANTIZED_H3_L4',
      'QUANTIZED_H3_L5',
      'QUANTIZED_H3_L6',
      'QUANTIZED_H3_L7',
      'QUANTIZED_H3_L8',
      'QUANTIZED_H3_L9',
      'QUANTIZED_H3_L10',
      'QUANTIZED_H3_L11',
      'QUANTIZED_H3_L12',
      'QUANTIZED_H3_L13',
      'QUANTIZED_H3_L14',
      'QUANTIZED_H3_L15',
      'QUANTIZED_TIMESTAMP_SECOND',
      'QUANTIZED_TIMESTAMP_MINUTE',
      'QUANTIZED_TIMESTAMP_HOUR',
      'QUANTIZED_TIMESTAMP_DAY',
      'QUANTIZED_TIMESTAMP_MONTH',
      'QUANTIZED_TIMESTAMP_YEAR',
    ]),
    'filter_ratio': 1.0,
    'filtered_count': 100,
    'filtered_rows_count': 100,
    'geo_projection': dict({
      'current_bucketing_column': None,
      'h3_level': None,
      'is_aggregated': False,
      'max_rows': 100000,
      'projection_rows': 100,
      'target_column': None,
    }),
    'has_data': True,
    'memory_usage': '0.15 MB',
    'row_count': 100,
    'shape': list([
      100,
      33,
    ]),
    'status': 'loaded',
    'temporal_projection': dict({
      'aggregation_level': None,
      'current_bucketing_column': None,
      'is_aggregated': False,
      'max_rows': 100000,
      'projection_rows': 100,
      'target_column': None,
    }),
  })
# ---
